Prepare the rest of WebCore for making the AtomString(const String&) constructor explicit

cdumez opened the pull request:
#### 636e305863afafacb8e2ee1f80e0a173e36762a1
<pre>
Prepare the rest of WebCore for making the AtomString(const String&amp;) constructor explicit
<a href="https://bugs.webkit.org/show_bug.cgi?id=239917">https://bugs.webkit.org/show_bug.cgi?id=239917</a>

Reviewed by Darin Adler.

Prepare the rest of WebCore for making the AtomString(const String&amp;) constructor explicit.
This helps find suboptimal patterns in our code base, where we may be doing unnecessary
String allocations.

* Source/WebCore/PAL/pal/FileSizeFormatter.cpp:
(PAL::fileSizeDescription):
* Source/WebCore/PAL/pal/FileSizeFormatter.h:
* Source/WebCore/PAL/pal/cocoa/FileSizeFormatterCocoa.mm:
(PAL::fileSizeDescription):
* Source/WebCore/Modules/airplay/WebKitPlaybackTargetAvailabilityEvent.idl:
* Source/WebCore/Modules/encryptedmedia/MediaKeyMessageEvent.idl:
* Source/WebCore/Modules/encryptedmedia/MediaKeySession.idl:
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeyMessageEvent.idl:
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeyNeededEvent.idl:
* Source/WebCore/Modules/mediarecorder/MediaRecorderErrorEvent.idl:
* Source/WebCore/Modules/mediastream/MediaStreamTrackEvent.idl:
* Source/WebCore/Modules/mediastream/OverconstrainedErrorEvent.idl:
* Source/WebCore/Modules/mediastream/RTCErrorEvent.idl:
* Source/WebCore/Modules/mediastream/RTCPeerConnectionIceErrorEvent.idl:
* Source/WebCore/Modules/mediastream/RTCPeerConnectionIceEvent.idl:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformErrorEvent.idl:
* Source/WebCore/Modules/mediastream/RTCTrackEvent.idl:
* Source/WebCore/Modules/mediastream/RTCTransformEvent.idl:
* Source/WebCore/Modules/notifications/NotificationEvent.idl:
* Source/WebCore/Modules/paymentrequest/MerchantValidationEvent.idl:
* Source/WebCore/Modules/paymentrequest/PaymentMethodChangeEvent.idl:
* Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.idl:
* Source/WebCore/Modules/speech/SpeechRecognitionErrorEvent.idl:
* Source/WebCore/Modules/speech/SpeechRecognitionEvent.idl:
* Source/WebCore/Modules/speech/SpeechSynthesisEvent.idl:
* Source/WebCore/Modules/webxr/XRInputSourceEvent.idl:
* Source/WebCore/Modules/webxr/XRInputSourcesChangeEvent.idl:
* Source/WebCore/Modules/webxr/XRReferenceSpaceEvent.idl:
* Source/WebCore/Modules/webxr/XRSessionEvent.idl:
* Source/WebCore/bindings/IDLTypes.h:
(WebCore::IDLString::isNullValue):
* Source/WebCore/bindings/js/JSDOMConvertStrings.h:
(WebCore::propertyNameToString):
(WebCore::Converter&lt;IDLLegacyNullToEmptyAtomStringAdaptor&lt;T &gt; &gt;::convert):
(WebCore::JSConverter&lt;IDLLegacyNullToEmptyAtomStringAdaptor&lt;T &gt; &gt;::convert):
(WebCore::JSConverter&lt;IDLAtomStringAdaptor&lt;T &gt; &gt;::convert):
(WebCore::JSConverter&lt;IDLAtomStringAdaptor&lt;IDLUSVString &gt; &gt;::convert):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateAttributeGetterBodyDefinition):
(GetAnnotatedIDLType):
* Source/WebCore/bindings/scripts/test/TestPromiseRejectionEvent.idl:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue):
(WebCore::CSSComputedStyleDeclaration::getPropertyValue):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/DOMCSSCustomPropertyDescriptor.h:
* Source/WebCore/css/DOMCSSCustomPropertyDescriptor.idl:
* Source/WebCore/css/MediaQueryExpression.cpp:
(WebCore::consumeFirstValue):
* Source/WebCore/css/MediaQueryListEvent.idl:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::MutableStyleProperties::setCustomProperty):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseValueWithVariableReferences):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseTypedCustomPropertyValue):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFamilyNameRaw):
* Source/WebCore/dom/DOMImplementation.cpp:
(WebCore::DOMImplementation::createDocumentType):
(WebCore::DOMImplementation::createDocument):
* Source/WebCore/dom/DOMImplementation.h:
* Source/WebCore/dom/DOMImplementation.idl:
* Source/WebCore/dom/Document+HTMLObsolete.idl:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createElementNS):
(WebCore::Document::formElementsState const):
(WebCore::Document::setStateForNewFormElements):
(WebCore::Document::parseQualifiedName):
(WebCore::Document::createAttributeNS):
(WebCore::Document::getCachedLocale):
(WebCore::Document::setBgColor):
(WebCore::Document::setFgColor):
(WebCore::Document::setAlinkColor):
(WebCore::Document::setLinkColorForBindings):
(WebCore::Document::setVlinkColor):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::toggleAttribute):
(WebCore::Element::didAddAttribute):
(WebCore::Element::didModifyAttribute):
(WebCore::Element::didRemoveAttribute):
* Source/WebCore/dom/KeyboardEvent.idl:
* Source/WebCore/dom/MutationObserver.h:
* Source/WebCore/dom/MutationObserver.idl:
* Source/WebCore/dom/QualifiedName.h:
(WebCore::QualifiedName::toAtomString const):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestModuleScript):
* Source/WebCore/editing/ApplyStyleCommand.cpp:
(WebCore::joinWithSpace):
* Source/WebCore/editing/CompositeEditCommand.cpp:
(WebCore::CompositeEditCommand::inputEventTypeName const):
* Source/WebCore/editing/CompositeEditCommand.h:
* Source/WebCore/editing/CreateLinkCommand.cpp:
(WebCore::CreateLinkCommand::doApply):
* Source/WebCore/editing/EditCommand.cpp:
(WebCore::inputTypeNameForEditingAction):
* Source/WebCore/editing/EditCommand.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::StyleChange::extractTextStyles):
* Source/WebCore/editing/EditingStyle.h:
(WebCore::StyleChange::fontColor):
(WebCore::StyleChange::fontFace):
(WebCore::StyleChange::fontSize):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::applyStyle):
(WebCore::Editor::applyParagraphStyle):
(WebCore::Editor::insertAttachment):
(WebCore::Editor::styleForSelectionStart):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeFormatBlock):
(WebCore::executeInsertHorizontalRule):
(WebCore::executeInsertImage):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::shouldAddNamespaceElement):
(WebCore::MarkupAccumulator::generateUniquePrefix):
* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::TextManipulationController::replace):
* Source/WebCore/editing/TypingCommand.cpp:
(WebCore::TypingCommand::inputEventTypeName const):
* Source/WebCore/editing/TypingCommand.h:
* Source/WebCore/editing/WebContentReader.h:
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::createFragmentForImageAttachment):
(WebCore::replaceRichContentWithAttachments):
(WebCore::createFragmentAndAddResources):
(WebCore::sanitizeMarkupWithArchive):
(WebCore::attachmentForFilePath):
(WebCore::attachmentForData):
(WebCore::WebContentReader::readURL):
(WebCore::WebContentReader::readDataBuffer):
* Source/WebCore/editing/markup.cpp:
(WebCore::AttributeChange::AttributeChange):
(WebCore::completeURLs):
(WebCore::replaceSubresourceURLs):
(WebCore::StyledMarkupAccumulator::appendCustomAttributes):
(WebCore::createFragmentForImageAndURL):
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::setDocumentState):
(WebCore::HistoryItem::documentState const):
* Source/WebCore/history/HistoryItem.h:
* Source/WebCore/html/FTPDirectoryDocument.cpp:
(WebCore::FTPDirectoryDocumentParser::createTDForFilename):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::saveFormControlState const):
* Source/WebCore/html/FormController.cpp:
(WebCore::StringVectorReader::consumeString):
(WebCore::StringVectorReader::consumeSubvector):
(WebCore::appendSerializedFormControlState):
(WebCore::FormController::formElementsState const):
(WebCore::FormController::setStateForNewFormElements):
(WebCore::FormController::parseStateVector):
(WebCore::FormController::referencedFilePaths):
* Source/WebCore/html/FormController.h:
* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::handleClick):
* Source/WebCore/html/HTMLAreaElement.idl:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::setFile):
(WebCore::HTMLAttachmentElement::updateAttributes):
(WebCore::HTMLAttachmentElement::updateEnclosingImageWithData):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/HTMLBaseElement.idl:
* Source/WebCore/html/HTMLButtonElement.idl:
* Source/WebCore/html/HTMLCollection.cpp:
(WebCore::HTMLCollection::isSupportedPropertyName):
* Source/WebCore/html/HTMLCollection.h:
* Source/WebCore/html/HTMLFormControlElementWithState.h:
* Source/WebCore/html/HTMLFormControlsCollection.cpp:
(WebCore::HTMLFormControlsCollection::namedItemOrItems const):
* Source/WebCore/html/HTMLFormControlsCollection.h:
* Source/WebCore/html/HTMLFormElement.idl:
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::openURL):
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::shouldLoadFrameLazily):
* Source/WebCore/html/HTMLIFrameElement.idl:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebCore/html/HTMLMapElement.cpp:
(WebCore::HTMLMapElement::parseAttribute):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setPreload):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLMediaElement.idl:
* Source/WebCore/html/HTMLOrForeignElement.idl:
* Source/WebCore/html/HTMLSelectElement.idl:
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::saveFormControlState const):
* Source/WebCore/html/HTMLTextAreaElement.idl:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::HTMLVideoElement):
* Source/WebCore/html/HiddenInputType.cpp:
(WebCore::HiddenInputType::saveFormControlState const):
(WebCore::HiddenInputType::restoreFormControlState):
(WebCore::HiddenInputType::setValue):
* Source/WebCore/html/ImageDocument.cpp:
(WebCore::ImageDocument::createDocumentStructure):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::saveFormControlState const):
* Source/WebCore/html/MediaDocument.cpp:
(WebCore::MediaDocumentParser::createDocumentStructure):
(WebCore::MediaDocument::replaceMediaElementTimerFired):
* Source/WebCore/html/ModelDocument.cpp:
(WebCore::ModelDocumentParser::createDocumentStructure):
* Source/WebCore/html/PluginDocument.cpp:
(WebCore::PluginDocumentParser::createDocumentStructure):
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::createAutoFillButton):
(WebCore::TextFieldInputType::updateAutoFillButton):
* Source/WebCore/html/canvas/WebGLContextEvent.idl:
* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::HTMLMetaCharsetParser::processMeta):
(WebCore::HTMLMetaCharsetParser::encodingFromMetaAttributes):
* Source/WebCore/html/parser/HTMLMetaCharsetParser.h:
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes):
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::createForeignAttributesMap):
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::SearchFieldCancelButtonElement::create):
* Source/WebCore/html/track/AudioTrack.cpp:
(WebCore::AudioTrack::updateKindFromPrivate):
* Source/WebCore/html/track/LoadableTextTrack.cpp:
(WebCore::LoadableTextTrack::LoadableTextTrack):
(WebCore::LoadableTextTrack::create):
* Source/WebCore/html/track/LoadableTextTrack.h:
* Source/WebCore/html/track/TextTrackCue.cpp:
(WebCore::TextTrackCue::setId):
* Source/WebCore/html/track/TextTrackCue.h:
(WebCore::TextTrackCue::id const):
* Source/WebCore/html/track/TextTrackCue.idl:
* Source/WebCore/html/track/TrackEvent.idl:
* Source/WebCore/html/track/VideoTrack.cpp:
(WebCore::VideoTrack::updateKindFromPrivate):
* Source/WebCore/html/track/WebVTTParser.cpp:
(WebCore::WebVTTParser::collectWebVTTBlock):
(WebCore::WebVTTParser::collectCueId):
(WebCore::WebVTTParser::resetCueValues):
* Source/WebCore/html/track/WebVTTParser.h:
* Source/WebCore/inspector/DOMEditor.cpp:
(WebCore::DOMEditor::setAttribute):
(WebCore::DOMEditor::removeAttribute):
* Source/WebCore/inspector/DOMEditor.h:
* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
(WebCore::InspectorAuditAccessibilityObject::getComputedProperties):
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::drawRulers):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::InspectorStyleSheetForInlineStyle::setStyleText):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::setAttributeValue):
(WebCore::InspectorDOMAgent::setAttributesAsText):
(WebCore::InspectorDOMAgent::removeAttribute):
(WebCore::InspectorDOMAgent::setNodeName):
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::interceptWithResponse):
(WebCore::InspectorNetworkAgent::interceptRequestWithResponse):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::load):
(WebCore::FrameLoader::loadPostRequest):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::FrameLoader::HistoryController::recursiveSetProvisionalItem):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):
* Source/WebCore/loader/PolicyChecker.h:
* Source/WebCore/page/PageSerializer.cpp:
(WebCore::isCharsetSpecifyingNode):
* Source/WebCore/platform/Pasteboard.h:
(WebCore::PasteboardWebContentReader::readDataBuffer):
* Source/WebCore/platform/graphics/InbandGenericCue.h:
(WebCore::GenericCueData::GenericCueData):
(WebCore::GenericCueData::decode):
(WebCore::InbandGenericCue::id const):
(WebCore::InbandGenericCue::setId):
* Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.mm:
(WebCore::InbandChapterTrackPrivateAVFObjC::processChapters):
* Source/WebCore/platform/graphics/iso/ISOVTTCue.cpp:
(WebCore::ISOWebVTTCue::ISOWebVTTCue):
(WebCore::ISOWebVTTCue::parse):
* Source/WebCore/platform/graphics/iso/ISOVTTCue.h:
* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::Pasteboard::read):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::RealtimeVideoCaptureSource):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::AVVideoCaptureSource):
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.h:
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
(WebCore::MockRealtimeVideoSource::create):
(WebCore::MockRealtimeVideoSourceMac::createForMockDisplayCapturer):
(WebCore::MockRealtimeVideoSourceMac::MockRealtimeVideoSourceMac):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::MockDisplayCapturer::MockDisplayCapturer):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::create):
(WebCore::MockRealtimeVideoSource::MockRealtimeVideoSource):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertStringOrAutoAtom):
(WebCore::Style::BuilderConverter::convertStringOrNoneAtom):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::shadowPseudoId):
(WebCore::Internals::setShadowPseudoId):
(WebCore::Internals::formControlStateOfPreviousHistoryItem):
(WebCore::Internals::setFormControlStateOfPreviousHistoryItem):
(WebCore::Internals::openDummyInspectorFrontend):
(WebCore::Internals::highlightPseudoElementColor):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebCore/testing/MockCDMFactory.cpp:
(WebCore::MockCDMFactory::setSupportedRobustness): Deleted.
* Source/WebCore/testing/MockCDMFactory.h:
(WebCore::MockCDMFactory::setSupportedRobustness):
* Source/WebCore/testing/MockCDMFactory.idl:

Canonical link: <a href="https://commits.webkit.org/250160@main">https://commits.webkit.org/250160@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@293656">https://svn.webkit.org/repository/webkit/trunk@293656</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

darinadler commented on Source/WebCore/bindings/js/JSDOMConvertStrings.h:
```diff
@@ -170,6 +170,25 @@ template<typename T>  struct JSConverter<IDLLegacyNullToEmptyStringAdaptor<T>> {
+    }
+};
+
+template<typename T>  struct JSConverter<IDLLegacyNullToEmptyAtomStringAdaptor<T>> {
```
Extra space here before `struct`.

darinadler commented on Source/WebCore/editing/EditorCommand.cpp:
```diff
@@ -408,9 +408,12 @@ static bool executeForeColor(Frame& frame, Event*, EditorCommandSource source, c
+    if (lowercaseValue[0] == '<' && lowercaseValue[lowercaseValue.length() - 1] == '>')
+        tagName = StringView(lowercaseValue).substring(1, lowercaseValue.length() - 2).toAtomString();
+    else
+        tagName = AtomString { lowercaseValue };
```
Should this be `WTFMove(lowercaseValue)`? And if that case is not already optimized, it should be, to save a tiny bit of reference count churn.

Separately, I will note that we could fuse the “converting to ASCII lowercase” and “making the atom string” into one operation so we do not need to allocate a temporary string for the case where there’s already an atom string, but the value we have includes an uppercase ASCII character. If that case is common enough then I suppose it would be worth it.

darinadler commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -466,7 +466,7 @@ void MarkupAccumulator::generateUniquePrefix(QualifiedName& prefixedName, const
     do {
         // FIXME: We should create makeAtomString, which would be more efficient.
-        name = makeString("NS", ++m_prefixLevel);
+        name = makeAtomString("NS", ++m_prefixLevel);
```
You added `_s` above where switching from `makeString` to `makeAtomString` but not here. Should be consistent within the same patch. I suppose given the direction you have been going, eventually we will change `StringConcatenate` to require `ASCIILiteral` or `StringView` rather than having it work on `const char*`.

darinadler commented on Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
```diff
@@ -295,12 +295,12 @@ static void replaceRichContentWithAttachments(Frame& frame, DocumentFragment& fr
     for (auto& attachment : descendantsOfType<HTMLAttachmentElement>(fragment)) {
         auto resourceURL = HTMLAttachmentElement::archiveResourceURL(attachment.uniqueIdentifier());
-        auto resourceEntry = urlToResourceMap.find(resourceURL.string());
+        auto resourceEntry = urlToResourceMap.find(AtomString { resourceURL.string() });
```
Given this is a find of an existing `AtomString` key, it can use an “existing atom string” code path to avoid waste fully allocating any temporary strings. We have that in `StringView`, not sure if we have the same for `String`.

darinadler commented on Source/WebCore/html/FormController.cpp:
```diff
@@ -42,21 +42,21 @@ static HTMLFormElement* ownerForm(const HTMLFormControlElementWithState& control
 
 struct StringVectorReader {
-    const Vector<String>& vector;
+    const Vector<AtomString>& vector;
```
Seems like we should rename the class `AtomStringVectorReader`.

darinadler commented on Source/WebCore/html/FormController.cpp:
```diff
@@ -267,14 +267,14 @@ Vector<String> FormController::formElementsState(const Document& document) const
-        stateVector.append(formKey);
-        stateVector.append(String::number(i - formStart));
+        stateVector.append(AtomString { formKey });
+        stateVector.append(AtomString::number(i - formStart));
```
We should probably delete `AtomString::number` and `String::number` and just use `makeAtomString` and `makeString`, presuming we can approve they have identical or better efficiency when passed a single numeric argument.

darinadler commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -558,7 +558,7 @@ void HTMLInputElement::updateType()
 
     if (didStoreValue && !willStoreValue && hasDirtyValue()) {
-        setAttributeWithoutSynchronization(valueAttr, m_valueIfDirty);
+        setAttributeWithoutSynchronization(valueAttr, AtomString { m_valueIfDirty });
```
Should `m_valueIfDirty` be kept as an atom string all along?

zdobersek committed:
[RISCV64] Implement MacroAssembler::probe(), ctiMasmProbeTrampoline
https://bugs.webkit.org/show_bug.cgi?id=239938

Reviewed by Yusuke Suzuki.

Implement MacroAssembler::probe() for RISCV64, along with the
ctiMasmProbeTrampoline operation. The implementation follows the process
of implementations for other platforms, with incoming, outgoing and
return-address-restoration records used to store register values during
setup and breakdown of the probe.

Going into the probe and back out of it, the general-purpose and
floating-point registers are stored, with the exception of global and
thread registers (x3 and x4). After the probe, if the probe state on the
stack is broken, the complete state is re-established before the
registers are reloaded, with the stack pointer and return address
registers set up last.

Covered by probing-related unit tests in testmasm.

* assembler/MacroAssemblerRISCV64.cpp:
(JSC::MacroAssembler::probe):

Canonical link: https://commits.webkit.org/250158@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@293654 268f45cc-cd09-0410-ab3c-d52691b4dbfc

Commit:
[GLIB] Update test expectations. Unreviewed test gardening.
https://bugs.webkit.org/show_bug.cgi?id=239945

Patch by Arcady Goldmints-Orlov <agoldmints@igalia.com> on 2022-05-01

* platform/glib/TestExpectations:
* platform/gtk/TestExpectations:

Canonical link: https://commits.webkit.org/250159@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@293655 268f45cc-cd09-0410-ab3c-d52691b4dbfc

cdumez committed:
Prepare the rest of WebCore for making the AtomString(const String&) constructor explicit
https://bugs.webkit.org/show_bug.cgi?id=239917

Reviewed by Darin Adler.

Prepare the rest of WebCore for making the AtomString(const String&) constructor explicit.
This helps find suboptimal patterns in our code base, where we may be doing unnecessary
String allocations.

* Source/WebCore/PAL/pal/FileSizeFormatter.cpp:
(PAL::fileSizeDescription):
* Source/WebCore/PAL/pal/FileSizeFormatter.h:
* Source/WebCore/PAL/pal/cocoa/FileSizeFormatterCocoa.mm:
(PAL::fileSizeDescription):
* Source/WebCore/Modules/airplay/WebKitPlaybackTargetAvailabilityEvent.idl:
* Source/WebCore/Modules/encryptedmedia/MediaKeyMessageEvent.idl:
* Source/WebCore/Modules/encryptedmedia/MediaKeySession.idl:
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeyMessageEvent.idl:
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeyNeededEvent.idl:
* Source/WebCore/Modules/mediarecorder/MediaRecorderErrorEvent.idl:
* Source/WebCore/Modules/mediastream/MediaStreamTrackEvent.idl:
* Source/WebCore/Modules/mediastream/OverconstrainedErrorEvent.idl:
* Source/WebCore/Modules/mediastream/RTCErrorEvent.idl:
* Source/WebCore/Modules/mediastream/RTCPeerConnectionIceErrorEvent.idl:
* Source/WebCore/Modules/mediastream/RTCPeerConnectionIceEvent.idl:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformErrorEvent.idl:
* Source/WebCore/Modules/mediastream/RTCTrackEvent.idl:
* Source/WebCore/Modules/mediastream/RTCTransformEvent.idl:
* Source/WebCore/Modules/notifications/NotificationEvent.idl:
* Source/WebCore/Modules/paymentrequest/MerchantValidationEvent.idl:
* Source/WebCore/Modules/paymentrequest/PaymentMethodChangeEvent.idl:
* Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.idl:
* Source/WebCore/Modules/speech/SpeechRecognitionErrorEvent.idl:
* Source/WebCore/Modules/speech/SpeechRecognitionEvent.idl:
* Source/WebCore/Modules/speech/SpeechSynthesisEvent.idl:
* Source/WebCore/Modules/webxr/XRInputSourceEvent.idl:
* Source/WebCore/Modules/webxr/XRInputSourcesChangeEvent.idl:
* Source/WebCore/Modules/webxr/XRReferenceSpaceEvent.idl:
* Source/WebCore/Modules/webxr/XRSessionEvent.idl:
* Source/WebCore/bindings/IDLTypes.h:
(WebCore::IDLString::isNullValue):
* Source/WebCore/bindings/js/JSDOMConvertStrings.h:
(WebCore::propertyNameToString):
(WebCore::Converter<IDLLegacyNullToEmptyAtomStringAdaptor<T>>::convert):
(WebCore::JSConverter<IDLLegacyNullToEmptyAtomStringAdaptor<T>>::convert):
(WebCore::JSConverter<IDLAtomStringAdaptor<T>>::convert):
(WebCore::JSConverter<IDLAtomStringAdaptor<IDLUSVString>>::convert):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateAttributeGetterBodyDefinition):
(GetAnnotatedIDLType):
* Source/WebCore/bindings/scripts/test/TestPromiseRejectionEvent.idl:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue):
(WebCore::CSSComputedStyleDeclaration::getPropertyValue):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/DOMCSSCustomPropertyDescriptor.h:
* Source/WebCore/css/DOMCSSCustomPropertyDescriptor.idl:
* Source/WebCore/css/MediaQueryExpression.cpp:
(WebCore::consumeFirstValue):
* Source/WebCore/css/MediaQueryListEvent.idl:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::MutableStyleProperties::setCustomProperty):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseValueWithVariableReferences):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseTypedCustomPropertyValue):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFamilyNameRaw):
* Source/WebCore/dom/DOMImplementation.cpp:
(WebCore::DOMImplementation::createDocumentType):
(WebCore::DOMImplementation::createDocument):
* Source/WebCore/dom/DOMImplementation.h:
* Source/WebCore/dom/DOMImplementation.idl:
* Source/WebCore/dom/Document+HTMLObsolete.idl:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createElementNS):
(WebCore::Document::formElementsState const):
(WebCore::Document::setStateForNewFormElements):
(WebCore::Document::parseQualifiedName):
(WebCore::Document::createAttributeNS):
(WebCore::Document::getCachedLocale):
(WebCore::Document::setBgColor):
(WebCore::Document::setFgColor):
(WebCore::Document::setAlinkColor):
(WebCore::Document::setLinkColorForBindings):
(WebCore::Document::setVlinkColor):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::toggleAttribute):
(WebCore::Element::didAddAttribute):
(WebCore::Element::didModifyAttribute):
(WebCore::Element::didRemoveAttribute):
* Source/WebCore/dom/KeyboardEvent.idl:
* Source/WebCore/dom/MutationObserver.h:
* Source/WebCore/dom/MutationObserver.idl:
* Source/WebCore/dom/QualifiedName.h:
(WebCore::QualifiedName::toAtomString const):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestModuleScript):
* Source/WebCore/editing/ApplyStyleCommand.cpp:
(WebCore::joinWithSpace):
* Source/WebCore/editing/CompositeEditCommand.cpp:
(WebCore::CompositeEditCommand::inputEventTypeName const):
* Source/WebCore/editing/CompositeEditCommand.h:
* Source/WebCore/editing/CreateLinkCommand.cpp:
(WebCore::CreateLinkCommand::doApply):
* Source/WebCore/editing/EditCommand.cpp:
(WebCore::inputTypeNameForEditingAction):
* Source/WebCore/editing/EditCommand.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::StyleChange::extractTextStyles):
* Source/WebCore/editing/EditingStyle.h:
(WebCore::StyleChange::fontColor):
(WebCore::StyleChange::fontFace):
(WebCore::StyleChange::fontSize):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::applyStyle):
(WebCore::Editor::applyParagraphStyle):
(WebCore::Editor::insertAttachment):
(WebCore::Editor::styleForSelectionStart):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeFormatBlock):
(WebCore::executeInsertHorizontalRule):
(WebCore::executeInsertImage):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::shouldAddNamespaceElement):
(WebCore::MarkupAccumulator::generateUniquePrefix):
* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::TextManipulationController::replace):
* Source/WebCore/editing/TypingCommand.cpp:
(WebCore::TypingCommand::inputEventTypeName const):
* Source/WebCore/editing/TypingCommand.h:
* Source/WebCore/editing/WebContentReader.h:
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::createFragmentForImageAttachment):
(WebCore::replaceRichContentWithAttachments):
(WebCore::createFragmentAndAddResources):
(WebCore::sanitizeMarkupWithArchive):
(WebCore::attachmentForFilePath):
(WebCore::attachmentForData):
(WebCore::WebContentReader::readURL):
(WebCore::WebContentReader::readDataBuffer):
* Source/WebCore/editing/markup.cpp:
(WebCore::AttributeChange::AttributeChange):
(WebCore::completeURLs):
(WebCore::replaceSubresourceURLs):
(WebCore::StyledMarkupAccumulator::appendCustomAttributes):
(WebCore::createFragmentForImageAndURL):
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::setDocumentState):
(WebCore::HistoryItem::documentState const):
* Source/WebCore/history/HistoryItem.h:
* Source/WebCore/html/FTPDirectoryDocument.cpp:
(WebCore::FTPDirectoryDocumentParser::createTDForFilename):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::saveFormControlState const):
* Source/WebCore/html/FormController.cpp:
(WebCore::StringVectorReader::consumeString):
(WebCore::StringVectorReader::consumeSubvector):
(WebCore::appendSerializedFormControlState):
(WebCore::FormController::formElementsState const):
(WebCore::FormController::setStateForNewFormElements):
(WebCore::FormController::parseStateVector):
(WebCore::FormController::referencedFilePaths):
* Source/WebCore/html/FormController.h:
* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::handleClick):
* Source/WebCore/html/HTMLAreaElement.idl:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::setFile):
(WebCore::HTMLAttachmentElement::updateAttributes):
(WebCore::HTMLAttachmentElement::updateEnclosingImageWithData):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/HTMLBaseElement.idl:
* Source/WebCore/html/HTMLButtonElement.idl:
* Source/WebCore/html/HTMLCollection.cpp:
(WebCore::HTMLCollection::isSupportedPropertyName):
* Source/WebCore/html/HTMLCollection.h:
* Source/WebCore/html/HTMLFormControlElementWithState.h:
* Source/WebCore/html/HTMLFormControlsCollection.cpp:
(WebCore::HTMLFormControlsCollection::namedItemOrItems const):
* Source/WebCore/html/HTMLFormControlsCollection.h:
* Source/WebCore/html/HTMLFormElement.idl:
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::openURL):
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::shouldLoadFrameLazily):
* Source/WebCore/html/HTMLIFrameElement.idl:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebCore/html/HTMLMapElement.cpp:
(WebCore::HTMLMapElement::parseAttribute):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setPreload):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLMediaElement.idl:
* Source/WebCore/html/HTMLOrForeignElement.idl:
* Source/WebCore/html/HTMLSelectElement.idl:
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::saveFormControlState const):
* Source/WebCore/html/HTMLTextAreaElement.idl:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::HTMLVideoElement):
* Source/WebCore/html/HiddenInputType.cpp:
(WebCore::HiddenInputType::saveFormControlState const):
(WebCore::HiddenInputType::restoreFormControlState):
(WebCore::HiddenInputType::setValue):
* Source/WebCore/html/ImageDocument.cpp:
(WebCore::ImageDocument::createDocumentStructure):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::saveFormControlState const):
* Source/WebCore/html/MediaDocument.cpp:
(WebCore::MediaDocumentParser::createDocumentStructure):
(WebCore::MediaDocument::replaceMediaElementTimerFired):
* Source/WebCore/html/ModelDocument.cpp:
(WebCore::ModelDocumentParser::createDocumentStructure):
* Source/WebCore/html/PluginDocument.cpp:
(WebCore::PluginDocumentParser::createDocumentStructure):
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::createAutoFillButton):
(WebCore::TextFieldInputType::updateAutoFillButton):
* Source/WebCore/html/canvas/WebGLContextEvent.idl:
* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::HTMLMetaCharsetParser::processMeta):
(WebCore::HTMLMetaCharsetParser::encodingFromMetaAttributes):
* Source/WebCore/html/parser/HTMLMetaCharsetParser.h:
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes):
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::createForeignAttributesMap):
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::SearchFieldCancelButtonElement::create):
* Source/WebCore/html/track/AudioTrack.cpp:
(WebCore::AudioTrack::updateKindFromPrivate):
* Source/WebCore/html/track/LoadableTextTrack.cpp:
(WebCore::LoadableTextTrack::LoadableTextTrack):
(WebCore::LoadableTextTrack::create):
* Source/WebCore/html/track/LoadableTextTrack.h:
* Source/WebCore/html/track/TextTrackCue.cpp:
(WebCore::TextTrackCue::setId):
* Source/WebCore/html/track/TextTrackCue.h:
(WebCore::TextTrackCue::id const):
* Source/WebCore/html/track/TextTrackCue.idl:
* Source/WebCore/html/track/TrackEvent.idl:
* Source/WebCore/html/track/VideoTrack.cpp:
(WebCore::VideoTrack::updateKindFromPrivate):
* Source/WebCore/html/track/WebVTTParser.cpp:
(WebCore::WebVTTParser::collectWebVTTBlock):
(WebCore::WebVTTParser::collectCueId):
(WebCore::WebVTTParser::resetCueValues):
* Source/WebCore/html/track/WebVTTParser.h:
* Source/WebCore/inspector/DOMEditor.cpp:
(WebCore::DOMEditor::setAttribute):
(WebCore::DOMEditor::removeAttribute):
* Source/WebCore/inspector/DOMEditor.h:
* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
(WebCore::InspectorAuditAccessibilityObject::getComputedProperties):
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::drawRulers):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::InspectorStyleSheetForInlineStyle::setStyleText):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::setAttributeValue):
(WebCore::InspectorDOMAgent::setAttributesAsText):
(WebCore::InspectorDOMAgent::removeAttribute):
(WebCore::InspectorDOMAgent::setNodeName):
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::interceptWithResponse):
(WebCore::InspectorNetworkAgent::interceptRequestWithResponse):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::load):
(WebCore::FrameLoader::loadPostRequest):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::FrameLoader::HistoryController::recursiveSetProvisionalItem):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):
* Source/WebCore/loader/PolicyChecker.h:
* Source/WebCore/page/PageSerializer.cpp:
(WebCore::isCharsetSpecifyingNode):
* Source/WebCore/platform/Pasteboard.h:
(WebCore::PasteboardWebContentReader::readDataBuffer):
* Source/WebCore/platform/graphics/InbandGenericCue.h:
(WebCore::GenericCueData::GenericCueData):
(WebCore::GenericCueData::decode):
(WebCore::InbandGenericCue::id const):
(WebCore::InbandGenericCue::setId):
* Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.mm:
(WebCore::InbandChapterTrackPrivateAVFObjC::processChapters):
* Source/WebCore/platform/graphics/iso/ISOVTTCue.cpp:
(WebCore::ISOWebVTTCue::ISOWebVTTCue):
(WebCore::ISOWebVTTCue::parse):
* Source/WebCore/platform/graphics/iso/ISOVTTCue.h:
* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::Pasteboard::read):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::RealtimeVideoCaptureSource):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::AVVideoCaptureSource):
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.h:
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
(WebCore::MockRealtimeVideoSource::create):
(WebCore::MockRealtimeVideoSourceMac::createForMockDisplayCapturer):
(WebCore::MockRealtimeVideoSourceMac::MockRealtimeVideoSourceMac):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::MockDisplayCapturer::MockDisplayCapturer):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::create):
(WebCore::MockRealtimeVideoSource::MockRealtimeVideoSource):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertStringOrAutoAtom):
(WebCore::Style::BuilderConverter::convertStringOrNoneAtom):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::shadowPseudoId):
(WebCore::Internals::setShadowPseudoId):
(WebCore::Internals::formControlStateOfPreviousHistoryItem):
(WebCore::Internals::setFormControlStateOfPreviousHistoryItem):
(WebCore::Internals::openDummyInspectorFrontend):
(WebCore::Internals::highlightPseudoElementColor):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebCore/testing/MockCDMFactory.cpp:
(WebCore::MockCDMFactory::setSupportedRobustness): Deleted.
* Source/WebCore/testing/MockCDMFactory.h:
(WebCore::MockCDMFactory::setSupportedRobustness):
* Source/WebCore/testing/MockCDMFactory.idl:

Canonical link: https://commits.webkit.org/250160@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@293656 268f45cc-cd09-0410-ab3c-d52691b4dbfc
